research/table_extraction_using_block_detection/code/labelled_blocks_to_csv.py

```diff
@@ -0,0 +1,189 @@
+        
+    def extract_term(self, titles, term):
+        '''
+        Extract Numbers based on particular term if present.
+        '''
+        filtered_titles = titles[titles.text.str.lower().str.contains(term)]
+        if len(filtered_titles) < 1:
+            return None
+        term_text = filtered_titles.text.iloc[0]
+        return ' '.join(re.findall('\d+', term_text))
+    
+    def detect_term(self, titles, term):
+        '''
+        Detect a term in title
+        '''
+        filtered_titles = titles[titles.text.str.lower().str.contains(term)]
+        if len(filtered_titles) > 0:
+            return True
+        return False
+            
+    def write_to_csv(self):
+        block_features = self.get_features_with_rows_and_cols()
+        tables = []
+        for table_no in block_features.table.unique():
+            # Retain/Extract the following features for table joining
+            #    - Page No
+            #    - Table No
+            #    - Major Head
+            #    - Demand No (if present)
+            table_features = block_features[block_features.table == table_no]
+            titles = table_features[table_features.label == 'title']
+            demand_no = self.extract_term(titles, 'demand no')
+            major_head = self.extract_term(titles, 'major head')
+            head_of_account = self.extract_term(titles, 'head of account')
+            abstract = self.detect_term(titles, 'abstract')
+            detailed = self.detect_term(titles, 'detailed')
+            detailed_account_no = self.extract_term(titles, 'detailed account no')
+            filename = '{0}_{1}.csv'.format(self.page_num, table_no)
+            tables.append({'page_no': self.page_num, 
+                           'table': table_no,
+                           'demand_no': demand_no,
+                           'major_head': major_head,
+                           'head_of_account': head_of_account,
+                           'detailed': detailed,
+                           'abstract': abstract,
+                           'detailed_account_no': detailed_account_no,
+                           'filename': filename})
+            if pd.notnull(table_features.col_index.max()):
+                max_col = int(table_features.col_index.max()) + 1
+            else:
+                max_col = 0
+            with open(filename, 'w') as csv_file:
+                # for each row write a line
+                add_header_prefix = False
+                if 'header' not in table_features.label.unique():
+                    number_of_default_headers = len(self.DEFAULT_HEADERS.split(';')) - 1
+                    if 'number_values' in table_features.label.unique():
+                        headers_row = ';' * (max_col - number_of_default_headers)
+                        headers_row += self.DEFAULT_HEADERS
+                    else:
+                        headers_row = ';;' + self.DEFAULT_HEADERS
```

Comment:
Purpose of this being?

Reply:
Let's get rid of semi-colons and rely on either Pandas -> CSV or traditional CSV file